File: leetcode/310.minimum-height-trees.ts
// @leet start
function findMinHeightTrees(n: number, edges: number[][]): number[] {
  if (n === 1) {
    return [0];
  }

  // Transform into undirected graph
  const graph: number[][] = Array.from({ length: n }, () => []);

  for (const edge of edges) {
    graph[edge[0]].push(edge[1]);
    graph[edge[1]].push(edge[0]);
  }

  // Get all 1-degree nodes and do something similar to Kahn's
  // algorithm
  let curr: number[] = [];
  const degree: number[] = new Array(n);

  for (const [node, edges] of graph.entries()) {
    degree[node] = edges.length;

    if (degree[node] === 1) {
      curr.push(node);
    }
  }

  while (true) {
    const next: number[] = [];

    for (const node of curr) {
      degree[node]--;

      for (const adj of graph[node]) {
        degree[adj]--;

        if (degree[adj] === 1) {
          next.push(adj);
        }
      }
    }

    if (!next.length) {
      return curr;
    }

    curr = next;
  }
}
// @leet end

File: leetcode/310.minimum-height-trees.py
# @leet start
class Solution:
    def findMinHeightTrees(self, n: int, edges: list[list[int]]) -> list[int]:
        if n == 1:
            return [0]

        indegree = [0] * n
        adj_lists: list[list[int]] = [[] for _ in range(n)]

        for a, b in edges:
            indegree[a] += 1
            indegree[b] += 1

            adj_lists[a].append(b)
            adj_lists[b].append(a)

        layer = [nd for nd in range(n) if indegree[nd] == 1]

        while True:
            next_layer: list[int] = []

            for nd in layer:
                for adj_nd in adj_lists[nd]:
                    indegree[adj_nd] -= 1

                    if indegree[adj_nd] == 1:
                        next_layer.append(adj_nd)

            if not next_layer:
                return layer

            layer = next_layer


# @leet end
